jsonrpc/server.go

```diff
@@ -0,0 +1,283 @@
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"reflect"
+	"strings"
+)
+
+const (
+	InvalidJson    = -32700 // Invalid JSON was received by the server.
+	InvalidRequest = -32600 // The JSON sent is not a valid Request object.
+	MethodNotFound = -32601 // The method does not exist / is not available.
+	InvalidParams  = -32602 // Invalid method parameter(s).
+	InternalError  = -32603 // Internal JSON-RPC error.
+)
+
+type request struct {
+	Version string `json:"jsonrpc"`
+	Method  string `json:"method"`
+	Params  any    `json:"params"`
+	Id      any    `json:"id"`
+}
+
+type batch []json.RawMessage
+
+type response struct {
+	Version string `json:"jsonrpc"`
+	Result  any    `json:"result,omitempty"`
+	Error   *Error `json:"error,omitempty"`
+	Id      any    `json:"id"`
+}
+
+type Error struct {
+	Code    int    `json:"code"`
+	Message string `json:"message"`
+	Data    any    `json:"data,omitempty"`
+}
+
+func newRequest(reader io.Reader) (*request, error) {
+	req := new(request)
+	dec := json.NewDecoder(reader)
+	dec.UseNumber()
+	if err := dec.Decode(req); err != nil {
+		return nil, err
+	} else if err = req.isSane(); err != nil {
+		return nil, err
+	}
+	return req, nil
+}
+
+func (r *request) isSane() error {
+	if r.Version != "2.0" {
+		return errors.New(fmt.Sprintf("unsupported RPC request version"))
+	}
+	if len(r.Method) <= 0 {
+		return errors.New("no method specified")
+	}
+
+	if r.Params != nil {
+		paramType := reflect.TypeOf(r.Params)
+		if paramType.Kind() != reflect.Slice && paramType.Kind() != reflect.Map {
+			return errors.New("params should be an array or an object")
+		}
+	}
+
+	if r.Id != nil {
+		idType := reflect.TypeOf(r.Id)
+		floating := idType.Name() == "Number" && strings.Contains(r.Id.(json.Number).String(), ".")
+		if (idType.Kind() != reflect.String && idType.Name() != "Number") || floating {
+			return errors.New("id should be a string or an integer")
+		}
+	}
+
+	return nil
+}
+
+type Parameter struct {
+	Name     string
+	Optional bool
+}
+
+type Method struct {
+	Name    string
+	Params  []Parameter
+	Handler any
+}
+
+type Server struct {
+	methods map[string]Method
+}
+
+// NewServer instantiates a JSONRPC server
+func NewServer() *Server {
+	return &Server{
+		methods: make(map[string]Method),
+	}
+}
+
+// RegisterMethod verifies and creates and endpoint that server recognizes.
+//
+// - name is the method name
+// - handler is the function to be called when a request is received for the
+// associated method. It should have (any, error) as it's return type
+// - paramNames are the names of parameters in the order that they are expected
+// by the handler
+func (s *Server) RegisterMethod(name string, params []Parameter, handler any) error {
+	handlerT := reflect.TypeOf(handler)
+	if handlerT.Kind() != reflect.Func {
+		return errors.New("handler must be a function")
+	}
+	if len(params) > 0 && handlerT.NumIn() != len(params) {
+		return errors.New("number of function params and param names must match")
+	}
```

Comment:
parameters are optional so why are we ensuring `len(params) > 0`?

Reply:
We are not ensuring that. We are simply checking that the handler accepts that many arguments; only if user provides params while registering